sdk/python/src/timecraft/app/app.py

```diff
@@ -0,0 +1,228 @@
+    """
+    _app: "App"
+    _tid: str
+    _addr: str
+    _resp: Optional[TaskResponse]
+    _done: bool
+
+    def __init__(self, app, tid):
+        self._app = app
+        self._tid = tid
+        self._done = False
+
+    async def result(self) -> Optional[TaskResponse]:
+        """
+        Returns the result of the task, or None if the task is not done.
+        The result is represented by a timecraft.TaskResponse object.
+        """
+        await self._update()
+        return self._resp
+
+    async def is_done(self) -> bool:
+        """
+        Returns True if the task is done, False otherwise.
+        """
+        await self._update()
+        return self._done
+
+    async def _update(self):
+        # TOOD: move to TaskState
+        def _is_done(state) -> bool:
+            done = [TaskState.SUCCESS, TaskState.ERROR]
+            return state in done
+
+        if self._done:
+            return
+
+        tasks = self._app._client.lookup_tasks([self._tid])
+
+        if len(tasks) > 1:
+            raise Exception("too many tasks returned")
+        if len(tasks) == 0:
+            raise Exception(f"task {self._tid} not found")
+
+        task = tasks[0]
+        if _is_done(task.state):
+            self._done = True
+            self._app._client.discard_tasks([self._tid])
+
+        self._resp = task
+
+
+class Function:
+    """
+    A Function is the execution unit of a timecraft App. Behind the scene,
+    a function will be scheduled as a task by the Ring scheduler.
+    """
+    _func: Callable[..., Any]
+    _app: "App"
+    _name: str
+
+    def __init__(self, app: "App", name: str, f: Callable[..., Any]) -> None:
+        self._func = f
+        self._app = app
+        self._name = name
+
+    def __call__(self, *args):
+        return self._func(*args)
+
+    def call(self, *args, **kwargs) -> Promise:
+        """
+        Calls the function with the given arguments and returns a Promise.
+        """
+        return self._app._spawn_task(self._name, *args, **kwargs)
+
+
+class App:
+    """
+    App is the main class of the Timecraft Python engine. An App will be used
+    to register a set of functions to be scheduled by the Ring scheduler.
+    """
+    _name: Optional[str]
+    _client: Client
+
+    _start: Optional[Entrypoint]
+    _functions: Dict[str, Function]
+
+    _web: aiohttp.web.Application
+
+    _source: Optional[str]
+
+    def __init__(self) -> None:
+        self._tasks = {}
+        self._client = Client()
+        self._functions = {}
+        self._start = None
+
+        # TODO: move somewhere else
+        self._web = aiohttp.web.Application()
+        self._web.add_routes([aiohttp.web.post("/", self._handle)])
+
+        # TODO: this is not reliable
+        self._source = inspect.stack()[1].filename
+
+    async def _serve(self):
+        # TODO: handle failures
+        await aiohttp.web._run_app(app=self._web, host="0.0.0.0", port=3000, print=None)
```

Comment:
Need a way to signal this guy to shutdown or some sort of handler to stop the guest on the Timecraft side.